### PR TITLE
Revise RESP implementation

### PR DESCRIPTION
**Pull request by raimis**
- [x] Clean up
- [x] Refactor grid generation
- [x] Implement restraints
- [x] Add some "hard" molecule benchmarks
- [x] Update tests

Depend on https://github.com/Acellera/htmd-data/pull/21

### CODE REVIEW

**Comment by raimis** [comment]
Atomic charge method comparison:

```
Molecule: methanol
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
C       0.0319    0.1167    0.1122    0.1085    0.0891    0.0079   -0.0385   -0.0182
O      -0.3996   -0.5988   -0.3164   -0.3156   -0.3110   -0.2891   -0.2361   -0.0935
H       0.0527    0.0287    0.0041    0.0051    0.0101    0.0310    0.0381    0.0154
H       0.0527    0.0287    0.0041    0.0051    0.0101    0.0310    0.0381    0.0154
H       0.0527    0.0287    0.0041    0.0051    0.0101    0.0310    0.0381    0.0154
H       0.2096    0.3960    0.1919    0.1919    0.1915    0.1883    0.1603    0.0655
```
```
Molecule: hydroxylamine
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
O      -0.3196   -0.4403   -0.2504   -0.2502   -0.2487   -0.2349   -0.1581   -0.0429
N      -0.2189   -0.6633   -0.3763   -0.3758   -0.3720   -0.3375   -0.1714   -0.0260
H       0.2372    0.4230    0.2275    0.2273    0.2254    0.2084    0.1199    0.0122
H       0.1507    0.3398    0.1996    0.1994    0.1976    0.1820    0.1048    0.0284
H       0.1507    0.3398    0.1996    0.1994    0.1976    0.1820    0.1048    0.0284
```
```
Molecule: urea
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
O      -0.2494   -0.6391   -0.3700   -0.3660   -0.3372   -0.2290   -0.1645   -0.1372
C       0.3091    0.7261    0.6056    0.5926    0.4972    0.1384   -0.0776   -0.0738
N      -0.3518   -0.6420   -0.5906   -0.5813   -0.5135   -0.2559   -0.0797   -0.0291
N      -0.3518   -0.6420   -0.5906   -0.5813   -0.5135   -0.2559   -0.0797   -0.0291
H       0.1610    0.2990    0.2364    0.2340    0.2167    0.1506    0.1004    0.0673
H       0.1610    0.2990    0.2364    0.2340    0.2167    0.1506    0.1004    0.0673
H       0.1610    0.2990    0.2364    0.2340    0.2167    0.1506    0.1004    0.0673
H       0.1610    0.2990    0.2364    0.2340    0.2167    0.1506    0.1004    0.0673
```
```
Molecule: isobutanol
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
C       0.0563    0.1458    0.4275    0.4343    0.3991    0.1899   -0.0047   -0.0278
C      -0.0366   -0.1091   -0.2974   -0.3032   -0.2684   -0.1001   -0.0181   -0.0072
C      -0.0366   -0.1091   -0.2974   -0.3032   -0.2684   -0.1001   -0.0181   -0.0072
C      -0.0366   -0.1091   -0.2974   -0.3032   -0.2684   -0.1001   -0.0181   -0.0072
O      -0.3908   -0.5988   -0.3387   -0.3396   -0.3349   -0.2986   -0.2311   -0.1162
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.0259    0.0427    0.0701    0.0714    0.0632    0.0265    0.0151    0.0096
H       0.2108    0.3970    0.1726    0.1725    0.1724    0.1705    0.1541    0.0789
```
```
Molecule: methylaldoxime
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
C      -0.0260   -0.1374   -0.3238   -0.3077   -0.2745   -0.0907   -0.0070   -0.0057
C       0.0404    0.3542    0.2471    0.2403    0.2245    0.1239   -0.0018   -0.0255
N      -0.1024   -0.4482   -0.1897   -0.1888   -0.1858   -0.1628   -0.1147   -0.0624
O      -0.4111   -0.3625   -0.2518   -0.2504   -0.2471   -0.2248   -0.1583   -0.0500
H       0.0291    0.0530    0.0889    0.0850    0.0769    0.0332    0.0195    0.0178
H       0.0291    0.0530    0.0889    0.0850    0.0769    0.0332    0.0195    0.0178
H       0.0291    0.0530    0.0889    0.0850    0.0769    0.0332    0.0195    0.0178
H       0.0824   -0.0032    0.0198    0.0208    0.0234    0.0419    0.0638    0.0321
H       0.3295    0.4370    0.2317    0.2309    0.2288    0.2129    0.1595    0.0581
```
```
Molecule: dimethylsulphonamine
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
C       0.0267   -0.1718   -0.1675   -0.1549   -0.1140   -0.1230   -0.0190   -0.0053
S       0.0905    0.4528    0.1297    0.1274    0.1208    0.1181    0.0677   -0.0111
O      -0.2430   -0.5142   -0.2027   -0.2024   -0.2020   -0.1996   -0.1807   -0.1222
N      -0.2312   -0.6403   -0.0266   -0.0330   -0.0505   -0.0623   -0.0747   -0.0221
C      -0.0006    0.2071   -0.1958   -0.1843   -0.1541   -0.1211   -0.0361   -0.0073
C      -0.0006    0.2071   -0.1958   -0.1843   -0.1541   -0.1211   -0.0361   -0.0073
H       0.0384    0.0804    0.0685    0.0651    0.0541    0.0568    0.0285    0.0223
H       0.0384    0.0804    0.0685    0.0651    0.0541    0.0568    0.0285    0.0223
H       0.0384    0.0804    0.0685    0.0651    0.0541    0.0568    0.0285    0.0223
H       0.0405    0.0360    0.0755    0.0727    0.0653    0.0564    0.0322    0.0181
H       0.0405    0.0360    0.0755    0.0727    0.0653    0.0564    0.0322    0.0181
H       0.0405    0.0360    0.0755    0.0727    0.0653    0.0564    0.0322    0.0181
H       0.0405    0.0360    0.0755    0.0727    0.0653    0.0564    0.0322    0.0181
H       0.0405    0.0360    0.0755    0.0727    0.0653    0.0564    0.0322    0.0181
H       0.0405    0.0360    0.0755    0.0727    0.0653    0.0564    0.0322    0.0181
```
```
Molecule: trisulphate
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
O      -0.7247   -0.7802   -1.0035   -1.0035   -1.0031   -0.9991   -0.8418   -0.4459
S       0.2327    1.7643    1.2500    1.2500    1.2500    1.2500    0.7697   -0.0724
O      -0.2120   -0.7802   -1.0035   -1.0035   -1.0031   -0.9991   -0.8418   -0.4459
O      -0.2120   -0.7802   -1.0035   -1.0035   -1.0031   -0.9991   -0.8418   -0.4459
O      -0.1440   -0.7214    1.0303    1.0301    1.0282    1.0101    1.0128    0.6529
S       0.3774    0.8723    1.2500    1.2500    1.2500    1.2500    1.1242    0.2189
O      -0.1822   -0.6578   -0.5409   -0.5409   -0.5408   -0.5406   -0.4843   -0.1496
O      -0.1822   -0.6578   -0.5409   -0.5409   -0.5408   -0.5406   -0.4843   -0.1496
S       0.2286    1.6906    1.2500    1.2500    1.2500    1.2500    0.9258   -0.0847
O      -0.2238   -0.6495   -0.8960   -0.8960   -0.8958   -0.8938   -0.7796   -0.3592
O      -0.2238   -0.6495   -0.8960   -0.8960   -0.8958   -0.8938   -0.7796   -0.3592
O      -0.7339   -0.6495   -0.8960   -0.8960   -0.8958   -0.8938   -0.7796   -0.3592
```
```
Molecule: phenyldiphosphate
Atom Gasteiger AM1-BCC   ESP b -10 ESP b -5  ESP b -4  ESP b -3  ESP b -2  ESP b -1 
O      -0.5635   -0.4732   -0.6332   -0.5862   -0.4821   -0.5860   -0.3366   -0.2741
P       0.5817    1.5159    1.2500    1.0444    0.6186    1.0468    0.0837   -0.0413
O      -0.0320   -0.4732   -0.6332   -0.5862   -0.4821   -0.5860   -0.3366   -0.2741
O      -0.0320   -0.4732   -0.6332   -0.5862   -0.4821   -0.5860   -0.3366   -0.2741
O       0.1437   -0.6898   -0.1855   -0.0536    0.1043   -0.0593    0.2111    0.1495
P       0.7879    1.5608    0.6826    0.4407    0.2793    0.4452    0.1046    0.0799
O       0.0204   -0.8050   -0.2104   -0.1467   -0.1022   -0.1477   -0.0501   -0.0488
O       0.0204   -0.8050   -0.2104   -0.1467   -0.1022   -0.1477   -0.0501   -0.0488
O      -0.2296   -0.5065   -0.3027   -0.2202   -0.1518   -0.2163    0.0111    0.0171
C       0.1894    0.1781    0.2129    0.1521    0.0968    0.1481    0.0255    0.0287
C      -0.0138   -0.1560    0.0687    0.0691    0.0586    0.0667    0.0084   -0.0017
C      -0.0584   -0.1200   -0.2390   -0.2099   -0.1746   -0.2058   -0.1261   -0.0933
C      -0.0620   -0.1620    0.0038   -0.0298   -0.0644   -0.0329   -0.0904   -0.1241
C      -0.0584   -0.1200   -0.2390   -0.2099   -0.1746   -0.2058   -0.1261   -0.0933
C      -0.0138   -0.1560    0.0687    0.0691    0.0586    0.0667    0.0084   -0.0017
H       0.0664    0.1605    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000
H       0.0624    0.1230    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000
H       0.0623    0.1170    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000
H       0.0624    0.1230    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000
H       0.0664    0.1605    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000
```

**Comment by raimis** [comment]
@cuzzo87 could you look to the charges and say, which scheme make the most sense, please.

**Comment by stefdoerr** [comment]
Wouldn't that depend on what forcefield you are using?

**Comment by raimis** [comment]
According to @cuzzo87, none of these atomic charge methods are good for all cases. The best candidates are Gasteiger and ESP without restraint (a.k.a. `ESP b -10`).

`GAFF` recommends (http://ambermd.org/antechamber/gaff.html#charge) to use RESP (but don't specify the restraint factor) or AM1-BCC.

Thus, we stay with our ESP!

**Comment by raimis** [comment]
@stefdoerr could you review, please.

File: htmd/charge/esp.py [hunk 1]
```diff
@@ -3,30 +3,175 @@
...
+
+    def _filterPoints(self, points):
+
+        # Compute distance threshold for each atom
+        thresholds = [radiusByElement(element) for element in self._molecule.element]
+        thresholds = min(self._shell_factors) * np.array(thresholds)
+
+        # Detect the points further away for each atom than its threshold
+        distances = cdist(points, self._molecule.coords[:, :, 0])
+        is_valid = np.all(distances > thresholds, axis=1)
```
Comment (stefdoerr) [inline htmd/charge/esp.py]: Isn't this risky? When you distribute points on the first shell with some numerical noise you might end up removing those points. Maybe you should add a small delta to the thresholds?

**Comment by stefdoerr** [comment]
Could we have some comparison code for antechamber/gaussian?

File: htmd/charge/esp.py (hunk 1 above)
Comment (raimis) [reply htmd/charge/esp.py]: Good point! I need to fix this.

**Comment by raimis** [comment]
Regarding, comparison with Antechamber, there two problems:
  1. Need Gaussian
  2. Their algorithm isn't well described

**Comment by stefdoerr** [comment]
About 2. I guess it should not matter much. 1. I can't help :P 
But if we cannot test against some reference implementation it's a bit risky, no?

**Comment by raimis** [comment]
I'm checking against @cuzzo87 intuition... :D

**Comment by stefdoerr** [comment]
Can't comment on the algorithm as I don't know it and we can't really compare it according to what you say.
I personally don't like classes that have a single used method or don't change state (like MoleculeGrid now is), I just implement them as functions but that's matter of taste.
So feel free to merge.

**Comment by raimis** [comment]
`MoleculeGrid` has two methods. :) I cannot merge until you approve...

**Pull request merged**